Several enhancements to GuestMemory and VolatileSlice

jiangliu opened the pull request.

bonzini commented:
Looks great, but I would like to understand the actual performance benefit from "_Optimize fast path for GuestMemory::read/write()_", since it introduces a little more complexity. No objection at all about the other patches!

andreeaflorescu commented:
@lauralg is working on adding performance tests for vm-memory. Should we first have some performance tests in place to actually see what is the impact of the patches? I'll let @lauralg and @alexandruag talk more about our findings here.

alexandruag commented:
Hi! Like Andreea mentioned, Laura wrote some tests and we've been playing around with them using the code available [here](https://github.com/alexandruag/vm-memory-test). They are simple synthetic benchmarks, but paint an interesting initial picture to do some comparative analysis (btw, if you run them, there will be some graphs at the end somewhere in the `target` folder, for easier visualisation). FWIW, it looks like the crosvm implementation fares better in general, and suffers a bit more when accessing later regions (maybe because it does a linear search, while we're doing a binary one right now). The changes from this PR sometime do better than master and sometimes a bit worse (tests only cover some cases tho). Have you also ran some tests to showcase the effects/improvements brought by these changes?

It looks quite important to have established ways of detecting regressions and/or measuring improvements as changes to vm-memory continue to happen (the same goes for other crates as well). We'll start a discussion about it soon to gather feedback and ideas from ppl. Also, regarding vm-memory in particular, I think we should take a good look at the experiences we had so far, validate some of the previous assumptions, polish (simplify if possible) the interface, and then work on various optimizations in earnest. The interface part is particularly important, because we've began using it in other components that are being developed (i.e. Virtio stuff), and we should ensure we're happy with it while changes are still relatively straightforward to make.

jiangliu commented:
> Looks great, but I would like to understand the actual performance benefit from "_Optimize fast path for GuestMemory::read/write()_", since it introduces a little more complexity. No objection at all about the other patches!

So let's move "Optimize fast path for GuestMemory::read/write()" out of this MR, and open another dedicated one for it:)

alexandruag commented on src/guest_memory.rs:
```diff
@@ -513,10 +513,20 @@ pub trait GuestMemory {
     }
 
-    /// Returns the address plus the offset if it is present within the memory of the guest.
+    /// Returns the address plus the offset if the whole range is present within the memory of
```
This is actually a breaking change in terms of `checked_offset` semantics. If the new functionality is helpful, we should just create a new method (`checked_range` or `checked_range_offset` or smt along those lines).

jiangliu commented:
> > Looks great, but I would like to understand the actual performance benefit from "_Optimize fast path for GuestMemory::read/write()_", since it introduces a little more complexity. No objection at all about the other patches!
> 
> So let's move "Optimize fast path for GuestMemory::read/write()" out of this MR, and open another dedicated one for it:)

With the recently added benchmark cases, it turns out that the optimization doesn't help much, so I will remove this patch.

jiangliu commented on src/volatile_memory.rs:
```diff
@@ -368,9 +368,19 @@ impl<'a> VolatileSlice<'a> {
-        source.copy_to(buf)
+        // A fast path for u8/i8
+        if size_of::<T>() == 1 {
+            // It is safe because the pointers are range-checked when the slices are created,
```
we could do that with a following MR:)

bonzini replied:
Note that this is not a pointer though that is being converted to a slice, it's a `&mut [T] where T: ByteValued` being converted to `&mut [u8]`.

lauralt commented on benches/guest_memory.rs:
```diff
@@ -0,0 +1,31 @@
+// Copyright (C) 2020 Alibaba Cloud Computing. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
```
According to discussions in this [PR](https://github.com/rust-vmm/vm-memory/pull/27), shouldn't this be:

`// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause`?

jiangliu commented on benches/volatile.rs:
```diff
@@ -0,0 +1,76 @@
+    benchmark_volatile_copy_to_volatile_slice(c);
+}
+
+fn benchmark_volatile_copy_to_u8(c: &mut Criterion) {
```
thanks, fixed:)

jiangliu commented on benches/volatile.rs:
```diff
@@ -0,0 +1,76 @@
+fn benchmark_volatile_copy_to_u8(c: &mut Criterion) {
+    let mut a = [0xa5u8; 1024];
+    let a_ref = &mut a[..];
+    let v_ref = a_ref.get_slice(0, a_ref.len()).unwrap();
```
That will fail due to concurrent mutable and immutable borrows.

jiangliu commented on benches/volatile.rs:
```diff
@@ -0,0 +1,76 @@
+    let v_ref = a_ref.get_slice(0, a_ref.len()).unwrap();
+    let d = [0xa5u8; 1024];
+
+    // Check performance for read operations.
```
yes, an error by copy & paste

jiangliu commented on src/volatile_memory.rs:
```diff
@@ -38,6 +38,8 @@ use std::usize;
 
 use crate::{ByteValued, Bytes};
 
+use copy_slice_impl::copy_slice;
```
In case we could provide different implementations for different arch or libc on demand.

jiangliu commented on src/volatile_memory.rs:
```diff
@@ -400,12 +399,10 @@ impl<'a> VolatileSlice<'a> {
-            // FIXME: ... however, is it really okay to mix non-volatile
-            // operations such as copy with read_volatile and write_volatile?
-            copy(self.addr, slice.addr, min(self.size, slice.size));
+            copy_slice(slice.as_mut_slice(), self.as_slice());
```
sure, it's a good idea to separate atomic accesses from byte stream access.

jiangliu replied:
https://github.com/rust-vmm/vm-memory/issues/111

jiangliu commented on src/guest_memory.rs:
```diff
@@ -513,6 +513,14 @@ pub trait GuestMemory {
     }
 
+    /// Returns the GuestAddress(base + len) if the range [base, base + len) is valid.
+    fn check_range(&self, base: GuestAddress, len: usize) -> bool {
```
I think a bool return value should be OK, because the typical usages are to validate the address range of a data structure. For example, when accessing a virtio descriptor, we could access individual fields after validating the whole descriptor is valid.

alexandruag commented on src/volatile_memory.rs:
```diff
@@ -30,7 +30,6 @@ use std::fmt;
 use std::io::{self, Read, Write};
 use std::marker::PhantomData;
 use std::mem::{align_of, size_of};
-use std::ptr::copy;
```
Hmm, on second though, would you mind dropping this commit for now? I think the issue you opened ensures we keep tracking the discussion, and after looking at the performance implications some more I think we might end up replacing `copy_slice` with `copy` or something similar throughout if the difference is significant (subject to performance testing ofc) and after the guaranteed atomicity aspect is no longer a requirement.

jiangliu replied:
sure, i will skip it to avoid performance regression.

alexandruag reviewed (commented):
Looks good! Just left a single comment.

lauralt commented on benches/guest_memory.rs:
```diff
@@ -0,0 +1,31 @@
+}
+
+fn benchmark_find_region(c: &mut Criterion) {
+    let mut regions: Vec<(GuestAddress, usize)> = Vec::new();
+    for i in 0..REGIONS_COUNT {
+        regions.push((GuestAddress(i * REGION_SIZE), REGION_SIZE as usize));
+    }
+
+    let memory = GuestMemoryMmap::from_ranges(regions.as_slice()).unwrap();
```
nit: this seems to be duplicated in `benchmark_for_mmap()`. It would be nice to have a helper function for it, something like:
`
pub fn create_memory(regions_count: u64, regions_size: size: u64) -> GuestMemoryMmap {}
` 
This way it would be also a bit easier to spot what the benchmark does.

Also, any reason to not keep this file in `benches/mmap` (and eventually add `#![cfg(feature = "backend-mmap")]` to it)?

jiangliu replied:
Add a helper to reduce duplicated code.
I aimed to keep consistence between benchmark files and source files:)

lauralt commented on benches/volatile.rs:
```diff
@@ -0,0 +1,45 @@
+    let d_slice = d_ref.get_slice(0, d_ref.len()).unwrap();
+
+    c.bench_function("VolatileSlice::copy_to_volatile_slice", |b| {
+        b.iter(|| black_box(a_slice.copy_to_volatile_slice(d_slice.clone())))
```
is `clone` needed here?

jiangliu replied:
yes, it's redundant and removed.

alexandruag commented:
Looks good! I clicked update branch and it seems to have added a merge commit, sorry :( Also the changelog update should go into the `## [Unreleased]` version, which gets renamed to the latest version number before publishing. Can you please rebase `upstream/master` so that new commit goes away?

jiangliu commented:
> Looks good! I clicked update branch and it seems to have added a merge commit, sorry :( Also the changelog update should go into the `## [Unreleased]` version, which gets renamed to the latest version number before publishing. Can you please rebase `upstream/master` so that new commit goes away?

No problem, fixed and rebased to the latest master branch:)
